File: HanJaehee/programmers/42898/programmers42898.java
class Solution {
  public int solution(int m, int n, int[][] puddles) {
    int[][] map = new int[n][m];

    for (int[] puddle : puddles)
      map[puddle[1] - 1][puddle[0] - 1] = -1;

    map[0][0] = 1;

    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) {

        if(map[i][j] == -1) {
          map[i][j] = 0;
          continue;
        }

        if(i != 0) map[i][j] += map[i - 1][j] % 1000000007; // 이거 나머지 처리 안하면 효율성 못뚫음;
        if(j != 0) map[i][j] += map[i][j - 1] % 1000000007;
      }
    }

    return map[n - 1][m - 1] % 1000000007;
  }

}

File: swanious/programmers/42898/programmers_42898.py
def solution(m, n, puddles):
    mapp = [[1] * m for _ in range(n)]
    for x, y in puddles: # 첫행, 첫열에 함정이 있으면 그곳부터 끝까지 갈 수 없음(우측, 아래로만 움직일 수 있으므로)
        mapp[y-1][x-1] = 0
        if y-1 == 0:
            for k in range(x-1, m):
                mapp[0][k] = 0
        if x-1 == 0:
            for k in range(y-1, n):
                mapp[k][0] = 0
        
    for i in range(1, n):
        for j in range(1, m):
            if mapp[i][j] != 0:
                mapp[i][j] = mapp[i-1][j] + mapp[i][j-1]
    
    return mapp[n-1][m-1]  % 1000000007